### PR TITLE
SQLCipher on windows

### PR DESCRIPTION
**Pull request by teams2ua**
Generate one big sqlite.c from sqlcipher repo as described https://sqlite.org/src/doc/trunk/README.md
Link with our own crypto lib.

### CODE REVIEW

File: .circleci/build_lib.sh [hunk 1]
```diff
@@ -106,8 +106,6 @@ export POLLY_ROOT=`pwd`/toolchains/polly
 ###
 # Clean
 ###
-echo "=====>Cleaning SQLCipher"
```
Comment (phaazon) [inline .circleci/build_lib.sh]: @KhalilBellakrid why was that needed, again?

File: CMakeLists.txt [hunk 2]
```diff
@@ -20,21 +20,9 @@ list(APPEND INCLUDE_DIRECTORIES core/test/include/)
...
-# step of SQLCipher project (because C compiler's path contains spaces)
-if (NOT MSVC AND NOT MINGW)
-    set(SQLCIPHER TRUE CACHE BOOL "Enable SQLCipher encryption support")
-endif()
-
-if (SQLCIPHER)
-    add_definitions("-DSQLCIPHER")
-    add_definitions("-DSQLITE_HAS_CODEC")
-endif()
+add_definitions("-DSQLITE_HAS_CODEC")
```
Comment (phaazon) [inline CMakeLists.txt]: Please keep the `SQLCIPHER` variable so that we can disable it if we want to.

File: core/lib/CMakeLists.txt [hunk 3]
```diff
@@ -14,11 +14,10 @@ add_subdirectory(snappy)
 add_subdirectory(leveldb)
 
 add_subdirectory(spdlog)
-if (SQLCIPHER)
-    include(ProjectSQLiteCipher)
-else()
-    add_subdirectory(sqlite3)
-endif()
+add_subdirectory(sqlcipher)
```
Comment (phaazon) [inline core/lib/CMakeLists.txt]: Ditto.

File: core/lib/soci_sqlite3/CMakeLists.txt [hunk 4]
```diff
@@ -14,23 +14,5 @@ add_library(soci_sqlite3 STATIC
...
-            find_library(FOUNDATION Foundation)
-            target_link_libraries(soci_sqlite3 ${FOUNDATION})
-            find_library(SECURITY Security)
-            target_link_libraries(soci_sqlite3 ${SECURITY})
-        endif()
-    endif()
-else()
-    target_include_directories(soci_sqlite3 PUBLIC ../sqlite3 )
-endif()
+target_link_libraries(soci_sqlite3 PUBLIC ${SQLITE_LIB})
```
Comment (phaazon) [inline core/lib/soci_sqlite3/CMakeLists.txt]: Ditto.

File: core/test/database/CMakeLists.txt [hunk 5]
```diff
@@ -16,13 +16,7 @@ add_executable(ledger-core-database-tests main.cpp pool_tests.cpp query_filters_
...
 
 target_link_libraries(ledger-core-database-tests gtest gtest_main)
-if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
-    # We need this otherwise link fails (for Linux),
-    # this is needed since we build static library for tests,
-    # not needed if we link against dynamic
-    # TODO: needs more investigation
-    target_link_libraries(ledger-core-database-tests "${CMAKE_SOURCE_DIR}/core/lib/sqlcipher/lib/${CMAKE_STATIC_LIBRARY_PREFIX}sqlcipher${CMAKE_STATIC_LIBRARY_SUFFIX}")
-endif()
+
```
Comment (phaazon) [inline core/test/database/CMakeLists.txt]: Extra whitespace.

File: core/test/database/database_soci_proxy_tests.cpp [hunk 6]
```diff
@@ -80,6 +80,7 @@ class SociProxyBaseTest : public ::testing::Test {
         _backend = std::make_shared<ProxyBackend>(engine);
         _backend->enableQueryLogging(true);
         auto date = DateUtils::toJSON(std::chrono::system_clock::now());
+        std::remove(date.begin(), date.end(), ':');
```
Comment (phaazon) [inline core/test/database/database_soci_proxy_tests.cpp]: Why is that needed?

File: core/test/database/sqlcipher_tests.cpp [hunk 7]
```diff
@@ -63,16 +64,6 @@ TEST(SQLCipherTest, SanityCheck) {
                 soci::use(value_0), soci::use(value_1);
         session.close();
     }
-    {
```
Comment (phaazon) [inline core/test/database/sqlcipher_tests.cpp]: I don’t get why you removed that from the test suite.

File: core/test/database/sqlcipher_tests.cpp [hunk 8]
```diff
@@ -104,12 +95,43 @@ TEST(SQLCipherTest, SanityCheck) {
...
         session.close();
+    }
+}
 
-        session.open(*factory, parameters);
+// Preventing terminating in environments that don't properly define __cplusplus macros
+// TODO: remove this check after we migrate to VS2017 and provide /Zc:__cplusplus option 
+// https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/
+#if __cplusplus >= 201103L
+TEST(SQLCipherTest, ThrowIfWrongPassword) {
```
Comment (phaazon) [inline core/test/database/sqlcipher_tests.cpp]: Hm, I don’t know how to feel about that. Tests shouldn’t be compiler-dependent like that. Also, what is that CPP conditional expression for? I don’t get the part with VS2017.

File: CMakeLists.txt (hunk 2 above)
Comment (teams2ua) [reply CMakeLists.txt]: If we want to stop encrypt, all we need to do is remove `add_definitions("-DSQLITE_HAS_CODEC")`
then sqlcipher will act exactly like sqlite.

File: core/test/database/database_soci_proxy_tests.cpp (hunk 6 above)
Comment (teams2ua) [reply core/test/database/database_soci_proxy_tests.cpp]: ":" can't be part of file name on windows, so the test fails.

File: .circleci/build_lib.sh (hunk 1 above)
Comment (KhalilBellakrid) [reply .circleci/build_lib.sh]: It was needed because we were building this external project before in-source, so when building for another target, it was necessary to clean it, otherwise build would fail, but now I guess we are good to go

File: core/test/database/sqlcipher_tests.cpp (hunk 7 above)
Comment (teams2ua) [reply core/test/database/sqlcipher_tests.cpp]: It was not removed it was moved to separate test that can only be run on not broke compilers. Will elaborate inside the next comment.

File: core/lib/CMakeLists.txt (hunk 3 above)
Comment (teams2ua) [reply core/lib/CMakeLists.txt]: no more sqlite3 - we have other implementation of this functionality in sqlcipher.

File: core/lib/soci_sqlite3/CMakeLists.txt (hunk 4 above)
Comment (teams2ua) [reply core/lib/soci_sqlite3/CMakeLists.txt]: Ditto `¯\_(ツ)_/¯`

File: core/test/database/sqlcipher_tests.cpp (hunk 8 above)
Comment (teams2ua) [reply core/test/database/sqlcipher_tests.cpp]: soci_core throws exceptions on the destructors. Starting from C++11 throwing exception in destructor should terminate program, unless the destructor was marked as `noexcept(false)` this is what soci_core is doing:
```
#if __cplusplus >= 201103L
#define SOCI_ONCE_TEMP_TYPE_NOEXCEPT noexcept(false)
#else
#define SOCI_ONCE_TEMP_TYPE_NOEXCEPT
#endif

and latter 

class SOCI_DECL once_temp_type
{
public:
...    
    ~once_temp_type() SOCI_ONCE_TEMP_TYPE_NOEXCEPT;
....
```
According to the article, all Visual Studios up to VS2017 15.7 defining __cplusplus as 199711L 
In other words Live Desktop on Windows will terminate if someone will try open encrypted DB with wrong password, don't know what to do here... either use another db driver, or migrate completely to VS2017 ...tuft choose

File: CMakeLists.txt (hunk 2 above)
Comment (phaazon) [reply CMakeLists.txt]: Then just surround it with a proper variable as it was before. I don’t want to have to edit a file while `-D` on the `cmake` command line invocation can do it.

Comment (teams2ua) [reply CMakeLists.txt]: If this option is not set, Windows build of sqlcipher is failing.
